src/test/java/com/google/devtools/build/lib/exec/SpawnLogContextTestBase.java

```diff
@@ -412,17 +491,743 @@ public void testRunfilesDirectoryInput(@TestParameter DirContents dirContents) t
+    Artifact externalGenArtifact =
+        ActionsTestUtil.createArtifact(
+            externalOutputDir,
+            someRepoOtherPkg
+                .getPackagePath(siblingRepositoryLayout)
+                .getChild("gen.txt")
+                .getPathString());
+    writeFile(externalGenArtifact, "external_gen");
+
+    Artifact symlinkSourceTarget = ActionsTestUtil.createArtifact(rootDir, "pkg/target.txt");
+    writeFile(symlinkSourceTarget, "symlink_source");
+    Artifact symlinkGenTarget = ActionsTestUtil.createArtifact(outputDir, "pkg/target.txt");
+    writeFile(symlinkGenTarget, "symlink_gen");
+
+    Artifact rootSymlinkSourceTarget =
+        ActionsTestUtil.createArtifact(rootDir, "pkg/root_target.txt");
+    writeFile(rootSymlinkSourceTarget, "root_symlink_source");
+    Artifact rootSymlinkGenTarget =
+        ActionsTestUtil.createArtifact(outputDir, "pkg/root_target.txt");
+    writeFile(rootSymlinkGenTarget, "root_symlink_gen");
+
+    Artifact runfilesMiddleman = ActionsTestUtil.createArtifact(middlemanDir, "runfiles");
+
+    PathFragment runfilesRoot = outputDir.getExecPath().getRelative("tools/foo.runfiles");
+    RunfilesTree runfilesTree =
+        createRunfilesTree(
+            runfilesRoot,
+            ImmutableMap.of(
+                "some/symlink", symlinkSourceTarget,
+                "other/symlink", symlinkGenTarget),
+            ImmutableMap.of(
+                "root/symlink", rootSymlinkSourceTarget,
+                "root/other/symlink", rootSymlinkGenTarget),
+            legacyExternalRunfiles,
+            sourceArtifact,
+            genArtifact,
+            externalSourceArtifact,
+            externalGenArtifact);
+
+    Spawn spawn = defaultSpawnBuilder().withInput(runfilesMiddleman).build();
+
+    SpawnLogContext context = createSpawnLogContext();
+
+    FakeActionInputFileCache inputMetadataProvider = new FakeActionInputFileCache();
+    inputMetadataProvider.putRunfilesTree(runfilesMiddleman, runfilesTree);
+    inputMetadataProvider.put(sourceArtifact, FileArtifactValue.createForTesting(sourceArtifact));
+    inputMetadataProvider.put(genArtifact, FileArtifactValue.createForTesting(genArtifact));
+    inputMetadataProvider.put(
+        externalSourceArtifact, FileArtifactValue.createForTesting(externalSourceArtifact));
+    inputMetadataProvider.put(
+        externalGenArtifact, FileArtifactValue.createForTesting(externalGenArtifact));
+    inputMetadataProvider.put(
+        symlinkSourceTarget, FileArtifactValue.createForTesting(symlinkSourceTarget));
+    inputMetadataProvider.put(
+        symlinkGenTarget, FileArtifactValue.createForTesting(symlinkGenTarget));
+    inputMetadataProvider.put(
+        rootSymlinkSourceTarget, FileArtifactValue.createForTesting(rootSymlinkSourceTarget));
+    inputMetadataProvider.put(
+        rootSymlinkGenTarget, FileArtifactValue.createForTesting(rootSymlinkGenTarget));
```

Comment:
Can we instead add a `createInputMetadataProvider` overload taking a `RunfilesTree` in addition to the list of artifacts and call it here? (Also for other test cases)